File: apps/sdkwork-rtc-pc/packages/sdkwork-rtc-pc-admin-core/src/services/roomService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type { Room, RoomCreateCommand, RoomListParams, RoomListResponse } from "../types/room";
import { resolveBackendRtcClient, type RtcBackendClientOptions, type RtcBackendClientSource } from "./backendClient";

export class RoomService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async list(params?: RoomListParams): Promise<RoomListResponse> {
    const response = await this.client.rtcRooms.rtc.rooms.list({
      page: params?.page,
      pageSize: params?.limit,
      cursor: params?.cursor,
      q: params?.search,
      sort: params?.sort,
      status: params?.status,
      ownerUserId: params?.ownerUserId,
      createdAfter: params?.createdAfter,
    });
    // The generated SDK unwraps the sdkwork-v3 envelope, so the response is
    // already the `data` payload ({ items, pageInfo }).
    const nextCursor = response.pageInfo?.nextCursor;
    return {
      items: response.items,
      nextCursor: nextCursor && nextCursor.length > 0 ? nextCursor : undefined,
    };
  }

  async get(id: string): Promise<Room> {
    const response = await this.client.rtcRooms.rtc.rooms.retrieve(id);
    if (!response) {
      throw new Error(`RTC room not found: ${id}`);
    }
    return response;
  }

  async create(command: RoomCreateCommand): Promise<Room> {
    const response = await this.client.rtcRooms.rtc.rooms.create({
      title: command.title,
      roomId: command.roomId ?? null,
    });
    return response;
  }
}

File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_admin_core/lib/src/services/room_service.dart
import '../admin_sdk_mapper.dart';
import '../backend_rtc_client.dart';
import '../models/room.dart';

class RoomListResult {
  final List<Room> items;
  final String? nextCursor;

  const RoomListResult({
    required this.items,
    this.nextCursor,
  });
}

class RoomService {
  final SdkworkBackendClient _client;

  RoomService(this._client);

  Future<RoomListResult> list({
    int? page,
    int? limit,
    String? cursor,
    String? search,
    String? sort,
    String? status,
    String? ownerUserId,
    String? createdAfter,
  }) async {
    final response = await _client.rtcRooms.list(
      page,
      limit,
      cursor,
      sort,
      search,
      status,
      ownerUserId,
      createdAfter,
    );
    return RoomListResult(
      items: backendResponseItems(response).map(Room.fromJson).toList(),
      nextCursor: backendResponseNextCursor(response),
    );
  }

  Future<Room> get(String id) async {
    final response = await _client.rtcRooms.retrieve(id);
    return Room.fromJson(
      backendResponseEntity(response, 'Room $id was not found'),
    );
  }
}
